src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,225 @@
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/OrderBookIsNotCrossed.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/InternalLedgerEntry.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "util/XDROperators.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include <fmt/format.h>
+#include <functional>
+#include <xdrpp/printer.h>
+
+namespace stellar
+{
+
+namespace
+{
+static AssetPairSet
+extractAssetPairs(LedgerTxnDelta const& ltxd)
+{
+    AssetPairSet assets;
+    for (auto const& entry : ltxd.entry)
+    {
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER && entry.second.current)
+        {
+            auto const& offer =
+                entry.second.current->ledgerEntry().data.offer();
+            auto const assetPair =
+                offer.selling < offer.buying
+                    ? std::make_pair(offer.selling, offer.buying)
+                    : std::make_pair(offer.buying, offer.selling);
+
+            assets.insert(assetPair);
+        }
+    }
+
+    return assets;
+}
+
+static double
+price(OfferEntry const& offer)
+{
+    return double(offer.price.n) / double(offer.price.d);
+}
+
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
+    // order book cannot be crossed
+    auto const iterAB = orderBook.find({a, b});
+    auto const iterBA = orderBook.find({b, a});
+    if (iterAB == orderBook.end() || iterBA == orderBook.end())
+    {
+        return {};
+    }
+
+    auto const& asks = iterAB->second;
+    auto const& bids = iterBA->second;
+    if (asks.empty() || bids.empty())
+    {
+        return {};
+    }
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
+
+    // Allow highestBid to equal (not exceed) lowestAsk if at least one of the
+    // offers is passive.
+    if (lowestAsk == highestBid)
+    {
+        auto existsEqualPriceNonPassive =
+            [](auto const targetPrice, Orders const& offers, auto& found) {
+                for (auto const& offer : offers)
+                {
+                    if (price(offer) != targetPrice)
+                    {
+                        return false;
+                    }
+                    if ((offer.flags & PASSIVE_FLAG) == 0)
+                    {
+                        found = offer;
+                        return true;
+                    }
+                }
+                return false;
+            };
+
+        OfferEntry ask, bid;
+        auto nonPassiveAskAtSamePrice =
+            existsEqualPriceNonPassive(lowestAsk, asks, ask);
```

Comment:
Here's what I think is true: there is an ask with price `lowestAsk`, but it could be passive, and if it is, then it doesn't indicate by itself that the order book is crossed (or perhaps it's more accurate to say, crossed in a way that we don't intend?  We intend, as I understand it, for a passive ask/bid to be able to be placed into the book even if there is an existing bid/ask at the same price).  However, if there is any other ask at the same price which isn't passive, and a bid at the same price which isn't passive, then we know that some non-passive ask/bid arrived at a time when another bid/ask at the same price was already present, so the new, non-passive offer should take taken the existing one; we must have a bug.  Is that right?

Reply:
Okay this works. Comment highly recommended because that is very strange looking.

Reply:
In light of discussion below, the implementation can be made faster.  The replacement will also need comments since it's a bit tricky too.  :)